pkg/destroy/vsphere/virtualmachines.go

```diff
@@ -0,0 +1,128 @@
+package vsphere
+
+import (
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+func (o *ClusterUninstaller) getVirtualMachineManagedObjects(moRef []types.ManagedObjectReference) ([]mo.VirtualMachine, error) {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	var virtualMachineMoList []mo.VirtualMachine
+	pc := property.DefaultCollector(o.Client)
+	err := pc.Retrieve(ctx, moRef, nil, &virtualMachineMoList)
+	if err != nil {
+		return nil, err
+	}
+	return virtualMachineMoList, nil
+}
+
+func (o *ClusterUninstaller) listVirtualMachines() ([]mo.VirtualMachine, error) {
+	tagAttachedObjects, err := o.getAttachedObjectsOnTag()
+	if err != nil && !isNotFound(err) {
+		return nil, err
+	}
+
+	// Seperate the objects attached to the tag based on type
+	// We only need VirtualMachine
+	var virtualMachineList []types.ManagedObjectReference
+	for _, attachedObject := range tagAttachedObjects {
+		for _, ref := range attachedObject.ObjectIDs {
+			if ref.Reference().Type == "VirtualMachine" {
+				virtualMachineList = append(virtualMachineList, ref.Reference())
+			}
+		}
+	}
+
+	var virtualMachineMoList []mo.VirtualMachine
+	if len(virtualMachineList) > 0 {
+		virtualMachineMoList, err = o.getVirtualMachineManagedObjects(virtualMachineList)
+	}
+
+	return virtualMachineMoList, err
+}
+
+func isPoweredOff(vmMO mo.VirtualMachine) bool {
+	return vmMO.Summary.Runtime.PowerState == "poweredOff"
+}
+
+func (o *ClusterUninstaller) stopVirtualMachine(vmMO mo.VirtualMachine) error {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	virtualMachineLogger := o.Logger.WithField("VirtualMachine", vmMO.Name)
+	if !isPoweredOff(vmMO) {
+		vm := object.NewVirtualMachine(o.Client, vmMO.Reference())
+		task, err := vm.PowerOff(ctx)
+		if err == nil {
+			err = task.Wait(ctx)
+		}
+		if err != nil {
+			virtualMachineLogger.Debug(err)
+			return err
+		}
+	}
+	virtualMachineLogger.Debug("Powered off")
+	return nil
+}
+
+func (o *ClusterUninstaller) stopVirtualMachines() error {
+	o.Logger.Debug("Power Off Virtual Machines")
+	found, err := o.listVirtualMachines()
+	if err != nil {
+		o.Logger.Debug(err)
+		return err
+	}
+
+	var errs []error
+	for _, vmMO := range found {
+		if !isPoweredOff(vmMO) {
+			if err := o.stopVirtualMachine(vmMO); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+func (o *ClusterUninstaller) deleteVirtualMachine(vmMO mo.VirtualMachine) error {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	virtualMachineLogger := o.Logger.WithField("VirtualMachine", vmMO.Name)
+	vm := object.NewVirtualMachine(o.Client, vmMO.Reference())
+	task, err := vm.Destroy(ctx)
+	if err == nil {
+		err = task.Wait(ctx)
```

Comment:
Do you know if the `Destroy` waits for the VM to actually be destroyed or does it just wait for the VM to be put into a destroying state? If it is the former, then should we try to parallelize the destroys?

Reply:
It won't block if you don't wait for the task and I am unsure how otherwise you would get the status - there is probably a way just not finding. it.

https://vdc-repo.vmware.com/vmwb-repository/dcr-public/6b586ed2-655c-49d9-9029-bc416323cb22/fa0b429a-a695-4c11-b7d2-2cbc284049dc/doc/vim.ManagedEntity.html#destroy